plenum/server/future_primaries_batch_handler.py

```diff
@@ -1,30 +1,75 @@
         super().__init__(database_manager, POOL_LEDGER_ID)
+        # map of view_no and list of primaries
+        self.primaries = {}    # type: Dict[int, List]
+        self.db_manager = database_manager
         self.node = node
 
+    def set_primaries(self, view_no, ps):
+        self.primaries[view_no] = ps
+
+    def _inspect_audit_txn(self, txn, view_no) -> int:
+        # Return delta of next primary's record
+        p_view_no = get_payload_data(txn)[AUDIT_TXN_VIEW_NO]
+        p_primaries = get_payload_data(txn)[AUDIT_TXN_PRIMARIES]
+        if p_view_no != view_no:
+            if isinstance(p_primaries, int):
+                return p_primaries
+            return 1
+        else:
+            if isinstance(p_primaries, int):
+                return p_primaries
+            return p_primaries
+
+    def _find_in_uncommitted(self, audit: Ledger, view_no, last_ind=1):
+        if len(audit.uncommittedTxns) == 0 or last_ind > len(audit.uncommittedTxns):
+            return None
+        last_txn = audit.uncommittedTxns[-last_ind]
+        primaries = self._inspect_audit_txn(last_txn, view_no)
+        if isinstance(primaries, list):
+            return primaries
+        return self._find_in_uncommitted(audit, view_no, last_ind + primaries)
+
+    def _get_previous_primaries(self, audit, view_no, seq_no_end):
+        if seq_no_end == 0:
+            return None
+        previous_txn = audit.getBySeqNo(seq_no_end)
+        primaries = self._inspect_audit_txn(previous_txn, view_no)
+        if isinstance(primaries, list):
+            return primaries
+        return self._get_previous_primaries(audit, view_no, seq_no_end - primaries)
+
+    def get_primaries_from_audit(self, view_no):
+        audit_ledger = self.db_manager.get_ledger(AUDIT_LEDGER_ID)
+        p_uncommited = self._find_in_uncommitted(audit_ledger, view_no)
+        if p_uncommited:
+            return p_uncommited
+        return self._get_previous_primaries(audit_ledger, view_no, audit_ledger.size)
+
     def post_batch_applied(self, three_pc_batch: ThreePcBatch, prev_handler_result=None):
-        node_reg = list(self.node.nodeReg.keys())
-        number_of_inst = getMaxFailures(len(node_reg)) + 1
         view_no = self.node.viewNo if three_pc_batch.original_view_no is None else three_pc_batch.original_view_no
-        validators = TxnPoolManager.calc_node_names_ordered_by_rank(node_reg, copy.deepcopy(self.node.nodeIds))
-        three_pc_batch.primaries = self.node.primaries_selector.select_primaries(view_no=view_no,
-                                                                                 instance_count=number_of_inst,
-                                                                                 validators=validators)
+        if view_no not in self.primaries:
+            ps = self.get_primaries_from_audit(view_no)
+            if not ps:
+                raise LogicError("Cannot restore primaries from Audit Ledger")
```

Comment:
It looks like we may need to use primaries from PrePrepare if we didn't find them in the audit ledger. We can trust them during re-ordering as we already have a quorum that the Primary is correct